### PR TITLE
SRIOV - syncVMI try to plug an SRIOV interface that is not in the pod

### PR DESCRIPTION
**Pull request by AlonaKaplan**
**What this PR does / why we need it**:
In case SRIOV interface/s exist, syncVMI tries to plug them to the VM.
If it fails, the whole syncVMI fails.
However, there are use cases we konw in advance it is fine to have a
SRIOV interface but not yet manage to plug it.
For example, when hotplugging a SRIOV interface. Until a migration is
done the interface/device is not plugged to the pod, therefore,
syncVM` will fail.
The PR allows ignoring SRIOV interface which are not yet plugged to
the pod and avoiding syncVMI failure.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/virt-launcher/virtwrap/device/hostdevice/sriov/hostdev.go [hunk 1]
```diff
@@ -39,8 +39,16 @@ import (
...
 func CreateHostDevices(vmi *v1.VirtualMachineInstance) ([]api.HostDevice, error) {
-	SRIOVInterfaces := vmispec.FilterSRIOVInterfaces(vmi.Spec.Domain.Devices.Interfaces)
+	SRIOVInterfaces := vmispec.FilterInterfacesSpec(vmi.Spec.Domain.Devices.Interfaces, func(iface v1.Interface) bool {
+		if iface.SRIOV == nil {
+			return false
+		}
+		ifaceStatus := vmispec.LookupInterfaceStatusByName(vmi.Status.Interfaces, iface.Name)
+		return ifaceStatus != nil && vmispec.ContainsInfoSource(ifaceStatus.InfoSource, vmispec.InfoSourceMultusStatus)
+	})
+	fmt.Printf("Alona ifaces %v\n", SRIOVInterfaces)
```
Comment (EdDev) [inline pkg/virt-launcher/virtwrap/device/hostdevice/sriov/hostdev.go:49]: Leftover

File: pkg/virt-launcher/virtwrap/device/hostdevice/sriov/hostdev.go [hunk 2]
```diff
@@ -39,8 +39,16 @@ import (
...
 	if len(SRIOVInterfaces) == 0 {
+		fmt.Printf("Alona no ifaces")
```
Comment (EdDev) [inline pkg/virt-launcher/virtwrap/device/hostdevice/sriov/hostdev.go:51]: Leftover

File: pkg/virt-launcher/virtwrap/manager_test.go [hunk 3]
```diff
@@ -32,6 +32,8 @@ import (
 	"strings"
 	"time"
 
+	"kubevirt.io/kubevirt/pkg/network/vmispec"
```
Comment (EdDev) [inline pkg/virt-launcher/virtwrap/manager_test.go:35]: nit: You can add it to the group below.

**Review by EdDev — COMMENTED** [review]
All looks good except some leftovers.

Also, please improve the commit messages with reasoning.

**Comment by ormergi** [comment]
@AlonaKaplan in what scenario it is known in advance that SRIOV hotplug will fail and result in sync error?

File: pkg/virt-launcher/virtwrap/device/hostdevice/sriov/hostdev.go [hunk 4]
```diff
@@ -39,8 +39,16 @@ import (
 )
 
 func CreateHostDevices(vmi *v1.VirtualMachineInstance) ([]api.HostDevice, error) {
-	SRIOVInterfaces := vmispec.FilterSRIOVInterfaces(vmi.Spec.Domain.Devices.Interfaces)
+	SRIOVInterfaces := vmispec.FilterInterfacesSpec(vmi.Spec.Domain.Devices.Interfaces, func(iface v1.Interface) bool {
+		if iface.SRIOV == nil {
+			return false
+		}
+		ifaceStatus := vmispec.LookupInterfaceStatusByName(vmi.Status.Interfaces, iface.Name)
+		return ifaceStatus != nil && vmispec.ContainsInfoSource(ifaceStatus.InfoSource, vmispec.InfoSourceMultusStatus)
```
Comment (ormergi) [inline pkg/virt-launcher/virtwrap/device/hostdevice/sriov/hostdev.go:47]: Up until now we kept virt-launcher from knowing those details (look in vmi status, compare it with spec..) and only compare between the domain and desired vmi spec.
Similar check already done at virt-handler https://github.com/kubevirt/kubevirt/blob/0379bbe4fd7e10fc6f9d1a96abe81e7bffafe5f2/pkg/virt-handler/vm.go#L3156-L3159
Note that there is also ratelimiting on SR-IOV hotplug operation.

Can we avoid doing this check in virt-launcher as well?

**Review by ormergi — COMMENTED** [review]
Please add reasoning for the first commit message

Regarding second commit message, there is typo s/hanged/handed/.

**Review by EdDev — APPROVED** [review]
Thanks!

File: pkg/virt-launcher/virtwrap/device/hostdevice/sriov/hostdev.go (hunk 4 above)
Comment (ormergi) [reply pkg/virt-launcher/virtwrap/device/hostdevice/sriov/hostdev.go:47]: I dont have better solution at the moment, once I have I will post a follow up PR.

**Review by ormergi — APPROVED** [review]
Thank you

**Comment by AlonaKaplan** [comment]
/approve

Raising two lgtms to approve

**Pull request merged**